tensorflow/core/common_runtime/dml/dml_execution_context.h

```diff
@@ -125,78 +132,96 @@ class DmlExecutionContextImpl {
-    return impl_->FillBufferWithPattern(dst, dst_offset, dst_size_in_bytes,
-                                        value);
-  }
-
-  DmlGpuEvent InitializeOperator(IDMLOperatorInitializer* initializer,
-                                 IDMLBindingTable* binding_table,
-                                 ID3D12DescriptorHeap* descriptor_heap) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return impl_->InitializeOperator(initializer, binding_table,
-                                     descriptor_heap);
-  }
-
-  DmlGpuEvent ExecuteOperator(IDMLCompiledOperator* op,
-                              IDMLBindingTable* binding_table,
-                              ID3D12DescriptorHeap* descriptor_heap) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return impl_->ExecuteOperator(op, binding_table, descriptor_heap);
-  }
+                                    absl::Span<const uint8_t> value);
+
+  // NOTE: the caller is responsible for keeping the initializer and
+  // descriptor_heap alive until the returned GPU event has completed.
+  DmlGpuEvent InitializeOperator(
+      IDMLOperatorInitializer* initializer,
+      Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+      ID3D12DescriptorHeap* descriptor_heap);
+
+  // NOTE: the caller is responsible for keeping the op and descriptor_heap
+  // alive until the returned GPU event has completed.
+  DmlGpuEvent ExecuteOperator(
+      IDMLCompiledOperator* op,
+      Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+      ID3D12DescriptorHeap* descriptor_heap);
 
   DmlGpuEvent ResourceBarrier(
-      absl::Span<const D3D12_RESOURCE_BARRIER> barriers) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return impl_->ResourceBarrier(barriers);
-  }
+      absl::Span<const D3D12_RESOURCE_BARRIER> barriers);
 
-  StatusOr<DmlGpuEvent> Flush() {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return impl_->Flush();
-  }
+  // A slightly more efficient version of ResourceBarrier when the barrier span
+  // only includes a UAV barrier (elides an extra copy).
+  DmlGpuEvent UavBarrier();
+
+  StatusOr<DmlGpuEvent> Flush();
 
   Status GetCommandRecorderStatus() const {
-    return impl_->GetCommandRecorderStatus();
+    return shared_state_->impl->GetCommandRecorderStatus();
   }
 
-  DmlGpuEvent GetCurrentCompletionEvent() {
-    std::unique_lock<std::mutex> lock(mutex_);
-    return impl_->GetCurrentCompletionEvent();
-  }
+  DmlGpuEvent GetCurrentCompletionEvent();
 
   D3D12_COMMAND_LIST_TYPE GetCommandListTypeForQueue() const {
-    return impl_->GetCommandListTypeForQueue();
+    return shared_state_->impl->GetCommandListTypeForQueue();
   }
 
  private:
-  std::mutex mutex_;
-  std::unique_ptr<DmlExecutionContextImpl> impl_;
+  static constexpr uint32_t default_batch_flush_size = 100;
+  static constexpr uint32_t default_batch_flush_time_us = 1000;
+
+  struct SharedState {
+    std::mutex mutex;
+    DmlGpuEvent next_flush_event;
```

Comment:
Is this always the same as the impl's CurrentCompletionEvent()? If not, under what circumstances will their fence values differ?

Reply:
I'm going to be refactoring this in a future change anyway, but it wouldn't be correct to call CurrentCompletionEvent here. Impl's version returns N if no ops have been recorded in the command list, or N+1 if they have. Since these methods are batching calls the impl will always return N, whereas we know these methods should return N+1 (they are, by definition, adding an op the command list).

Really though this separation between EC/ECImpl needs to be collaped.